packages/utils/telemetry-utils/src/config.ts

```diff
@@ -0,0 +1,263 @@
+            if (typeof v !== firstType) {
+                return defaultReturn;
+            }
+        }
+        return { ...defaultReturn, raw: input, [`${firstType}[]`]: output };
+    }
+
+    return defaultReturn;
+}
+
+const safeSessionStorage = (): Storage | undefined => {
+    try {
+        return sessionStorage !== null ? sessionStorage : undefined;
+    } catch { return undefined; }
+};
+
+/**
+ * Implementation of {@link IConfigProvider} which contains nested {@link IConfigProviderBase} instances
+ */
+export class CachedConfigProvider implements IConfigProvider {
+    private readonly configCache = new Map<string, StronglyTypedValue>();
+    private readonly orderedBaseProviders: (IConfigProviderBase | undefined)[];
+
+    constructor(
+        ... orderedBaseProviders: (IConfigProviderBase | undefined)[]
+    ) {
+        this.orderedBaseProviders = [];
+        const knownProviders = new Set<IConfigProviderBase>();
+        const candidateProviders = [...orderedBaseProviders];
+        while (candidateProviders.length > 0) {
+            const baseProvider = candidateProviders.shift()!;
+            if (baseProvider !== undefined
+                && isConfigProviderBase(baseProvider)
+                && !knownProviders.has(baseProvider)
+            ) {
+                knownProviders.add(baseProvider);
+                if (baseProvider instanceof CachedConfigProvider) {
+                    candidateProviders.push(...baseProvider.orderedBaseProviders);
+                } else {
+                    this.orderedBaseProviders.push(baseProvider);
+                }
+            }
+        }
+    }
+    getBoolean(name: string): boolean | undefined {
+        return this.getCacheEntry(name)?.boolean;
+    }
+    getNumber(name: string): number | undefined {
+        return this.getCacheEntry(name)?.number;
+    }
+    getString(name: string): string | undefined {
+        return this.getCacheEntry(name)?.string;
+    }
+    getBooleanArray(name: string): boolean[] | undefined {
+        return this.getCacheEntry(name)?.["boolean[]"];
+    }
+    getNumberArray(name: string): number[] | undefined {
+        return this.getCacheEntry(name)?.["number[]"];
+    }
+    getStringArray(name: string): string[] | undefined {
+        return this.getCacheEntry(name)?.["string[]"];
+    }
+
+    getRawConfig(name: string): ConfigTypes {
+        return this.getCacheEntry(name)?.raw;
+    }
+
+    private getCacheEntry(name: string): StronglyTypedValue | undefined {
+        if (!this.configCache.has(name)) {
+            for (const provider of this.orderedBaseProviders) {
+                const parsed = stronglyTypedParse(provider?.getRawConfig(name));
+                if (parsed !== undefined) {
+                    this.configCache.set(name, parsed);
+                    return parsed;
+                }
+            }
+            // configs are immutable, if the first lookup returned no results, all lookups should
```

Comment:
the comment references the behavior of this class, so you are correct

Reply:
Just giving feedback that the wording here was hard to understand at first.  Consider something like "cached config values are permanent" instead of "configs are immutable".

I'm bringing a different dialect to the table too, where "config" more naturally means what you're calling "config provider", and what you call "config" I would call "config value" or "setting".  So it's very subjective feedback :)